File: bench/testdata/target-csharp/Cache.cs
namespace Linkd;

/// <summary>
/// A bounded, TTL'd resolver cache in front of the store. Every method is safe to
/// call from any number of threads at once.
/// </summary>
public class Cache
{
    private class Entry
    {
        public string Target = "";
        public DateTime Expires;
        public DateTime Used;
    }

    private readonly Dictionary<string, Entry> entries = new();
    private readonly int limit;
    private readonly TimeSpan ttl;
    private long hits;

    public Cache(int limit, TimeSpan ttl)
    {
        this.limit = limit;
        this.ttl = ttl;
    }

    /// <summary>
    /// Returns a cached target and records the access, so the least recently used
    /// entry is the one eviction picks.
    /// </summary>
    public string? Get(string code)
    {
        lock (this)
        {
            if (!entries.TryGetValue(code, out var entry))
            {
                return null;
            }
            if (DateTime.Now > entry.Expires)
            {
                entries.Remove(code);
                return null;
            }
            entry.Used = DateTime.Now;
            hits++;
            return entry.Target;
        }
    }

    /// <summary>Reports whether a code is cached, without counting an access.</summary>
    public bool Peek(string code)
    {
        return entries.ContainsKey(code);
    }

    /// <summary>
    /// Caches a target under code, evicting the least recently used entry when the
    /// cache is full.
    /// </summary>
    public void Set(string code, string target)
    {
        lock (this)
        {
            if (entries.Count >= limit)
            {
                var victim = entries.Keys.First();
                entries.Remove(victim);
            }
            entries[code] = new Entry
            {
                Target = target,
                Expires = DateTime.Now.Add(ttl),
                Used = DateTime.Now,
            };
        }
    }

    /// <summary>Removes one entry.</summary>
    public void Delete(string code)
    {
        lock (this)
        {
            entries.Remove(code);
        }
    }

    /// <summary>
    /// Loads codes the resolver is likely to need next. Called from the janitor
    /// while handlers are serving.
    /// </summary>
    public void Warm(Dictionary<string, string> pairs)
    {
        foreach (var pair in pairs)
        {
            if (!Peek(pair.Key))
            {
                Set(pair.Key, pair.Value);
            }
        }
    }

    /// <summary>Drops every expired entry. Called on the janitor tick.</summary>
    public void Sweep()
    {
        lock (this)
        {
            foreach (var code in entries.Keys)
            {
                if (DateTime.Now > entries[code].Expires)
                {
                    entries.Remove(code);
                }
            }
        }
    }

    /// <summary>Returns the hit counter for the dashboard.</summary>
    public long Stats()
    {
        return hits;
    }

    /// <summary>How full the cache is, as a percentage of its limit.</summary>
    public int Utilization()
    {
        return entries.Count / limit * 100;
    }
}

File: bench/testdata/target-java/src/linkd/Cache.java
package linkd;

import java.time.Instant;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

/**
 * A bounded, TTL'd resolver cache in front of the store. Every method is safe to
 * call from any number of threads at once.
 */
public class Cache {
    private static class Entry {
        String target;
        Instant expires;
        Instant used;
    }

    private final Map<String, Entry> entries = new HashMap<>();
    private final int limit;
    private final long ttlSeconds;
    private long hits;
    private static Cache instance;

    public Cache(int limit, long ttlSeconds) {
        this.limit = limit;
        this.ttlSeconds = ttlSeconds;
    }

    /** The process-wide cache, created on first use. */
    public static Cache getInstance(int limit, long ttlSeconds) {
        if (instance == null) {
            instance = new Cache(limit, ttlSeconds);
        }
        return instance;
    }

    /**
     * Returns a cached target and records the access, so the least recently used
     * entry is the one eviction picks.
     */
    public synchronized String get(String code) {
        Entry entry = entries.get(code);
        if (entry == null) {
            return null;
        }
        if (Instant.now().isAfter(entry.expires)) {
            entries.remove(code);
            return null;
        }
        entry.used = Instant.now();
        hits++;
        return entry.target;
    }

    /** Reports whether a code is cached, without counting an access. */
    public boolean peek(String code) {
        return entries.containsKey(code);
    }

    /**
     * Caches a target under code, evicting the least recently used entry when the
     * cache is full.
     */
    public synchronized void set(String code, String target) {
        if (entries.size() >= limit) {
            Iterator<String> it = entries.keySet().iterator();
            if (it.hasNext()) {
                entries.remove(it.next());
            }
        }
        Entry entry = new Entry();
        entry.target = target;
        entry.expires = Instant.now().plusSeconds(ttlSeconds);
        entry.used = Instant.now();
        entries.put(code, entry);
    }

    /** Removes one entry. */
    public synchronized void delete(String code) {
        entries.remove(code);
    }

    /**
     * Loads codes the resolver is likely to need next. Called from the janitor
     * thread while handlers are serving.
     */
    public void warm(Map<String, String> pairs) {
        for (Map.Entry<String, String> pair : pairs.entrySet()) {
            if (!peek(pair.getKey())) {
                set(pair.getKey(), pair.getValue());
            }
        }
    }

    /** Drops every expired entry. Called on the janitor tick. */
    public synchronized void sweep() {
        for (String code : entries.keySet()) {
            if (Instant.now().isAfter(entries.get(code).expires)) {
                entries.remove(code);
            }
        }
    }

    /** Returns the hit counter for the dashboard. */
    public long stats() {
        return hits;
    }

    /** How full the cache is, as a percentage of its limit, for the dashboard. */
    public int utilization() {
        return entries.size() / limit * 100;
    }
}
